lib/model/recent_senders.dart

```diff
@@ -0,0 +1,133 @@
+import 'package:collection/collection.dart';
+import 'package:flutter/foundation.dart';
+
+import '../api/model/model.dart';
+
+class MessageIdTracker {
+  MessageIdTracker();
+
+  @visibleForTesting
+  MessageIdTracker.fromIds(List<int> ids) {
+    _ids.addAll([...{...ids}]..sort());
+  }
+
+  final List<int> _ids = [];
+
+  void add(int id) {
+    if (_ids.isEmpty) {
+      _ids.add(id);
+    } else {
+      int i = lowerBound(_ids, id);
+      if (i < _ids.length && _ids[i] == id) return; // the [id] already exists, so do not add it.
+      _ids.insert(i, id);
+    }
+  }
+
+  // TODO: remove
+
+  /// The maximum id in the tracker list.
+  ///
+  /// Returns -1 if the tracker list is empty.
+  int get maxId => _ids.isNotEmpty ? _ids.last : -1;
```

Comment:
Have we considered returning null instead of -1?

Reply:
Yeah, I think returning `null` is better as we preferred it in the previous PR (#693) too.